### PR TITLE
test(node): spend simulation test

### PR DESCRIPTION
**Pull request by RolandSherwin**
- attempts to perform spends / double spends / poison old spends and tries to verify all the states at the end.

### CODE REVIEW

File: .github/workflows/nightly_wan.yml [hunk 1]
```diff
@@ -1,8 +1,9 @@
 name: Nightly -- Full WAN Network Tests
 
 on:
-  schedule:
-    - cron: "0 0 * * *"
+# todo: this is totally broken atm. Fix and re-enable.
+#   schedule:
+#     - cron: "0 0 * * *"
```
Comment (grumbach) [inline .github/workflows/nightly_wan.yml:6]: Anyone we should tell this about? @jacderida?

File: sn_node/tests/spend_simulation.rs [hunk 2]
```diff
@@ -0,0 +1,697 @@
...
+            random_tx.hash()
+        );
+
+        return Ok(Some(random_tx.clone()));
+    }
+    Ok(None)
+}
+
+/// A spend / transaction is poisonable if all of its outputs are already spent.
+fn find_all_poisonable_spends<'a>(
```
Comment (grumbach) [inline sn_node/tests/spend_simulation.rs:671]: Actually, only one spent output is enough!

File: sn_node/tests/spend_simulation.rs [hunk 3]
```diff
@@ -0,0 +1,697 @@
...
+        received_cash_note: Vec<CashNote>,
+    },
+}
+
+#[derive(Debug)]
+enum SpendStatus {
+    Utxo,
+    Spent,
+    Poisoned,
+    // DoubleSpend,
```
Comment (maqi) [inline sn_node/tests/spend_simulation.rs]: shall this to be removed?

File: sn_node/tests/spend_simulation.rs [hunk 4]
```diff
@@ -0,0 +1,697 @@
...
+}
+
+#[derive(custom_debug::Debug)]
+struct State {
+    // immutable
+    #[debug(skip)]
+    action_senders: BTreeMap<usize, mpsc::Sender<WalletAction>>,
+    all_wallets: BTreeMap<usize, TempDir>,
+    main_pubkeys: BTreeMap<usize, MainPubkey>,
+    main_pubkeys_inverse: BTreeMap<MainPubkey, usize>,
```
Comment (maqi) [inline sn_node/tests/spend_simulation.rs]: I think some more detailed comments to explain what those parameters are for will be usefult.
for example, why need this extra `inverse` ?

File: sn_node/tests/spend_simulation.rs [hunk 5]
```diff
@@ -0,0 +1,697 @@
...
+}
+
+#[derive(Debug, Default)]
+struct PendingTasksTracker {
+    pending_send_results: Vec<usize>,
+    pending_receive_results: Vec<usize>,
+}
+
+#[tokio::test]
+async fn cash_note_transfer_double_spend_fail() -> Result<()> {
```
Comment (maqi) [inline sn_node/tests/spend_simulation.rs]: may need a more detailed comment to explain: what's the detailed scenary of the test, and the expected result .
some general description shall be enough.

File: sn_node/tests/spend_simulation.rs (hunk 2 above)
Comment (RolandSherwin) [reply sn_node/tests/spend_simulation.rs:671]: If an output is unspent, then that output cannot be spent later right. 
I wanted to address `poison`= even after a double spend, the entire branch is valid. i.e, any descendent can be spent.

File: sn_node/tests/spend_simulation.rs (hunk 4 above)
Comment (RolandSherwin) [reply sn_node/tests/spend_simulation.rs]: Will do that yeah!

**Review by grumbach — APPROVED** [review]
LGTM!

**Pull request merged**